blobstore.wit.md

```diff
@@ -0,0 +1,186 @@
+//
+// ## Object
+// An object is a named sequence of bytes. Objects can be read or written with a stream interface.
+// Once an object is written its contents cannot be modified.
+// Objects can be any size from 1 byte up to the limits imposed by the underlying store.
+// Object names are unique within their container.
+//
+// ## Container
+// A container is a named collection of objects. Container names can be any utf-8 string.
+// Within any instance of the blobstore interface, all containers share the same "flat" namespace.
+
+
+// wasi-cloud Blobstore service definition
+interface "wasi:blob/blobstore" {
+
+  use wasi:blob/types::{ Error, container-metadata, container-name, object-id }
+
+  // creates a new empty container
+  create-container: func(name: container-name) -> result<container,Error>
+
+  // get a list of all containers, as a stream.
+  list-containers: func() -> result<stream<container-name>, Error>
+
+  // retrieves a container by name
+  get-container: func(name: container-name) -> result<container, Error>
+
+  // deletes a container and all objects within it
+  delete-container: func(name: container-name) -> result<_, Error>
+
+  // returns true if the container exists
+  container-exists: func(name: container-name) -> result<bool, Error>
+
+  // copies (duplicates) an object, to the same or a different container.
+  // returns an error if the target container does not exist.
+  // overwrites destination object if it already existed.
+  copy-object: func(src: object-id, dest: object-id) -> result<_, Error>
+
+  // moves or renames an object, to the same or a different container
+  // returns an error if the destination container does not exist.
+  // overwrites destination object if it already existed.
+  move-object: func(src:object-id, dest: object-id) -> result<_, Error>
+}
+
+// a Container is a collection of objects
+resource container {
+
+  // returns container name
+  name: func() -> result<string, Error>
+
+  // returns container metadata
+  info: func() -> result<container-metadata, Error>
+
+  // begins reading an object
+  read-object: func(name: object-name) -> result<read-stream, Error>
+
+  // creates or replaces an object.
+  write-object: func(name: object-name) -> result<write-stream, Error>
+
+  // retrieves an object or portion of an object, as a resource.
+  // Start and end offsets are inclusive.
+  // Once a data-blob resource has been created, the underlying bytes are held by the blobstore service for the lifetime
+  // of the data-blob resource, even if the object they came from is later deleted.
+  get-data: func(name: object-name, start: u64, end: u64) -> result<data-blob, Error>
+
+  // creates or replaces an object with the data blob.
+  write-data: func(name: object-name, data: data-blob) -> result<_, Error>
+
+  // returns list of objects in the container. Order is undefined.
+  list-objects: func(name: object-name) -> result<stream<object-name>, Error>
```

Comment:
While this one is pretty common, it still may not be universally available

Reply:
It can also be a performance problem for some implementations if there are complicated access policies to evaluate for each object.  We don't want to require a method where an app might unknowingly invoke a very expensive operation on a server.

OTOH, there are some applications that need this functionality. Should it be moved to a separate interface?

As with byte-ranges, this seems like a good topic to raise in the community call.